### ISSUE TITLE
Required field in arrays

### ISSUE THREAD

**Issue by MatejJan** [peerlibrary/meteor-peerdb#44]
One of the things I struggle most is the required field in reference fields.

I think I understand the reasoning in normal cases, let's say with a post and comments. Comments will have a reference field to post and leave the default required to true. This means that if the post gets deleted the comment documents will get deleted too. This makes sense and I hope I have this correct.

But let's say you set it the other way around (for whatever reason). So this time post has a comments field that is an array reference field to comments. Because I don't want the post to be deleted if a comment gets deleted my intuition is to set required to false:

```
comments = [@ReferenceField Comment, ['text', 'author'], false]
```

But this throws an error: `Reference field directly in an array cannot be optional`.

How is this to be interpreted? My intuition goes from thinking about one comment. If I can only have one comment, I could set this to false. It would mean that if the one comment gets deleted, the comment reference here is set to null (instead of deleting the post). This works OK. But as soon as I want multiple comments, this doesn't hold any more?

The only way I can imagine a consistent interpretation would be that required in this case actually refers to the comment element inside the array - it is required and it will be deleted from the array if the comment gets deleted (as opposed to being replaced in a null placeholder in the array).

Is that true then?

So maybe this is more a suggestion for a better example section of required as it was definitely the hardest to understand as much as I remember my learning experience from the very beginning of my PeerDB use. Since it still boggles me sometimes after 2 or 3 years (like now), it might be a good place to put an extra example in or two.

**Comment by mitar** [comment]
No, post will not get deleted if comment in an array is deleted. Only object inside the array will be removed. This is why there is no point of having it optional: you do not want to have an array of comments to look like:

```
[{comment1}, null, {comment2}]
```

**Comment by mitar** [comment]
And sure, please make a pull request with an example. :-)

**Comment by MatejJan** [comment]
Yeah, so I guess I figured it out after an hour correct. :) Now that I think I have a grasp on it, I might just as well do that. I agree that `null`s don't have any place there. It just took a while to realized that's what required refers to.

Maybe it's not too off-topic to ask: why did you decide for `required` to be `true` by default?

**Comment by mitar** [comment]
Because to me `required` means that the sub-document has to be removed when where it is pointing to has been removed, instead of set to `null`. And in an array it should always be removed and never set to `null`.

**Comment by MatejJan** [comment]
But yeah, that's such a 'destructive' thing. That somewhere you're referencing, suddenly decides whether your document would get destroyed. So by default, any reference between two documents you create, it's a 'strong' link, child-parent relationship where child can't live without the parent.

I would think a 'weak' link would be more beginner-friendly. If you don't have a reference, that's how the object behaves. If you put a reference on, you just make sure they are actually objectified and can import some extra fields by default (if field is set). And then you have to decide and set if you want the reference to be strong (destructive to this document). So my reasoning is that if you had 'required' be 'false' by default, it's closer to behavior if reference was not set in the first place.

And I'm talking about outside of arrays.

For inside of arrays, I'm not super keen on required switching to mean the existence of the embedded document. Required between Comment->Post means that Comment will be removed if Post is removed. But when Post->[Comment] it's Post's Comment is removed if Comment is removed (not Post is removed if Comment is removed). I'd take it as a given, that if Comment is removed, that it will disappear everywhere else it's being referenced from. If the Comment is set to a field, sure, the comment 'disappears' and the field is left being set to null (assuming required=false). If the Comment was in an array, it 'disappears' as well and gets deleted from the array (array has one less element).

Maybe food for thought for PeerDB 2.0 (aka Meteor Documents ;)). It's just an alternative interpretation that is more intuitive for me (and reason why I struggled with how you have it defined and set by default).

But yeah, now that I finally know how it works, for me it doesn't matter, I'll be able to use it correct either way.

**Comment by mitar** [comment]
I agree, it is one of those things I thought of improving in Reactive Documents.

In DBs there are [multiple options](https://en.wikipedia.org/wiki/Propagation_constraint) here and I am hard-coding only one. It works in most cases, but I have also encountered that sometimes I would like a less strict version as well. I do not remember what what was the situation though. And I also think that later on I discovered that in is often just a signal of some strange schema you have, if you would like to keep invalid references around.
